shared/subspace-data-retrieval/src/object_fetcher/tests.rs

```diff
@@ -0,0 +1,829 @@
+            });
+
+        assert_eq!(bytes_written, object_len_encoded_len);
+    }
+}
+
+/// Creates a mapping from piece(s), start piece index, offset, and object length.
+/// Returns the mapping and object data.
+///
+/// If supplied:
+/// - `skip_padding` is the amount of padding to skip, and
+/// - `raw_data_after_segment_header` is the raw data from the piece at the start of the segment,
+///   after the segment header has been written to it.
+fn create_mapping(
+    pieces: Vec<&Piece>,
+    start_piece_index: usize,
+    offset: usize,
+    object_len: usize,
+    mut skip_padding: Option<usize>,
+    raw_data_after_segment_header: Option<Vec<u8>>,
+) -> (GlobalObject, Vec<u8>) {
+    // Simplify the code below by setting the default skip padding value.
+    if raw_data_after_segment_header.is_some() && skip_padding.is_none() {
+        skip_padding = Some(0);
+    }
+    if skip_padding.is_some() {
+        assert!(
+            raw_data_after_segment_header.is_some(),
+            "can only skip padding if there is more data in the next segment"
+        );
+    }
+
+    let object_len_encoded = compact_encoded(object_len);
+    let pieces_len = pieces.len();
+    let mut raw_data = extract_raw_data(pieces).collect::<Vec<u8>>();
+
+    if let (Some(skip_padding), Some(raw_data_after_segment_header)) =
+        (skip_padding, raw_data_after_segment_header)
+    {
+        // Find the next piece index with a segment header
+        let segment_header_piece_index =
+            start_piece_index.next_multiple_of(ArchivedHistorySegment::NUM_PIECES);
+        // And the piece before it can have padding
+        let padding_piece_index = segment_header_piece_index.checked_sub(1);
+
+        // Skip padding if needed
+        if let Some(padding_piece_index) = padding_piece_index
+            && skip_padding > 0
+        {
+            let original_len = raw_data.len();
+
+            let byte_position_in_raw_data = byte_position_in_extract_raw_data(
+                padding_piece_index,
+                start_piece_index,
+                pieces_len,
+                original_len,
+            );
+
+            // Delete the padding bytes we want to skip
+            let replaced_padding_data = raw_data
+                .splice(
+                    // The padding range we want to skip
+                    byte_position_in_raw_data + RawRecord::SIZE - skip_padding
+                        ..byte_position_in_raw_data + RawRecord::SIZE,
+                    // Delete the padding bytes
+                    iter::empty(),
+                )
+                .collect::<Vec<_>>();
+
+            assert_eq!(raw_data.len(), original_len - skip_padding);
+            assert_eq!(
+                replaced_padding_data,
+                vec![PADDING_BYTE_VALUE; skip_padding]
+            );
+        }
+
+        // If the offset hasn't already skipped the segment header, skip it now
+        if segment_header_piece_index > start_piece_index {
```

Comment:
Similarly I don't understand how this comment matches the condition. It only succeeds when `start_piece_index` belongs to a previous (completely different) segment, why comparing it with `segment_header_piece_index` belonging to the next segment?

Reply:
`next_multiple_of` is poorly named, it actually returns the multiple greater than or equal to the number. So this condition means "if the start piece is not the first piece in a segment". (If it is the first piece in a segment, then the offset will already be greater than the segment header, because blocks can only start after the header.)

I'll update this comment to explain that - and it might be clearer to use `!=` instead.

Reply:
I you want to check if the piece is the first in a segment, the most natural way to do that is to check if its position in a segment is `0`. The way it is written right now obfuscates the intention.

Reply:
To be fair, this is test code that I wrote in a bit of a hurry 🥲